staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/goopenapi.go

```diff
@@ -0,0 +1,144 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package schema
+
+import (
+	"github.com/go-openapi/spec"
+)
+
+// NewStructural converts a structural schema to go-openapi schema.
+func (s *Structural) ToGoOpenAPI() *spec.Schema {
+	if s == nil {
+		return nil
+	}
+
+	ret := &spec.Schema{}
+
+	if s.Items != nil {
+		ret.Items = &spec.SchemaOrArray{Schema: s.Items.ToGoOpenAPI()}
+	}
+	if s.Properties != nil {
+		ret.Properties = make(map[string]spec.Schema, len(s.Properties))
+		for k, v := range s.Properties {
+			ret.Properties[k] = *v.ToGoOpenAPI()
+		}
+	}
+	s.Generic.toGoOpenAPI(ret)
+	s.Extensions.toGoOpenAPI(ret)
+	s.ValueValidation.toGoOpenAPI(ret)
+
+	return ret
+}
+
+func (g *Generic) toGoOpenAPI(ret *spec.Schema) {
+	if g == nil {
+		return
+	}
+
+	if len(g.Type) != 0 {
+		ret.Type = spec.StringOrArray{g.Type}
+		if g.Nullable {
```

Comment:
why nested inside `if len(g.Type) != 0`?

Reply:
If the OpenAPI v3 type field is empty, nullable has no effect and we can just ignore it here. Actually, adding `type: null` would make the schema even stronger and wrong. In case of non-empty types, we add null to the type array: `type: ["object","null"|`.

I don't like that go-openapi has this JSON-schema-inspired type slice, but no nullable. The consequence is that JSON produced by go-openapi is either wrong (not OpenAPI v3 compliant), or we have to filter those multi-type values out. We do the latter in our v2 conversion.

After 1.15 I would like to rethink our use of go-openapi. What we really need is
1. OpenAPI v3 compatible types which can be marshalled to JSON (and preferribly also Protobuf to avoid many internal roundtrips)
2. An OpenAPI v3 compatible validator.

For (1) I would rather create our own types than using broken 3rd party libs.
For (2) we can keep using go-openapi, but very locally in CRD validation.

Reply:
For the v2 conversion, compare https://github.com/kubernetes/kubernetes/blob/d0c3b708026f681f7467d36dcf9696f89ccfcfab/staging/src/k8s.io/apiextensions-apiserver/pkg/controller/openapi/conversion.go#L66.